File: internal/history/history.go
// Package history is a lightweight store for past incidents: a directory of
// archived bundle files plus an index.json describing them. The Go side
// appends to it with `diagkit archive`; the Python analyzer reads it to list
// past incidents and find signatures that recur across them.
package history

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"

	"github.com/SAY-5/diagkit/internal/bundle"
)

// IndexVersion identifies the index layout. Both languages check it.
const IndexVersion = "1"

// IndexFile is the name of the index inside a history directory.
const IndexFile = "index.json"

// DefaultDir is where incidents land when no directory is given.
const DefaultDir = "diagkit-history"

// Entry summarizes one archived incident.
type Entry struct {
	ID         string        `json:"id"`
	File       string        `json:"file"`
	Scenario   string        `json:"scenario"`
	Seed       int64         `json:"seed"`
	Window     bundle.Window `json:"window"`
	Logs       int           `json:"logs"`
	Traces     int           `json:"traces"`
	Signatures int           `json:"signatures"`
}

// Index is the full history catalog.
type Index struct {
	Version   string  `json:"version"`
	Incidents []Entry `json:"incidents"`
}

// Load reads the index from dir. A missing directory or index yields an empty
// index, so the first archive call bootstraps the store.
func Load(dir string) (*Index, error) {
	data, err := os.ReadFile(filepath.Join(dir, IndexFile))
	if errors.Is(err, fs.ErrNotExist) {
		return &Index{Version: IndexVersion}, nil
	}
	if err != nil {
		return nil, err
	}
	var idx Index
	if err := json.Unmarshal(data, &idx); err != nil {
		return nil, fmt.Errorf("parse %s: %w", IndexFile, err)
	}
	if idx.Version != IndexVersion {
		return nil, fmt.Errorf("unsupported history index version %q, expected %q", idx.Version, IndexVersion)
	}
	return &idx, nil
}

// Archive copies the bundle into dir under the next sequential id and appends
// it to the index. The same incident archived twice becomes two entries; the
// recurrence report on the Python side is built on exactly that.
func Archive(dir string, b *bundle.Bundle) (Entry, error) {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return Entry{}, err
	}
	idx, err := Load(dir)
	if err != nil {
		return Entry{}, err
	}

	id := fmt.Sprintf("inc-%04d", len(idx.Incidents)+1)
	e := Entry{
		ID:         id,
		File:       id + ".json",
		Scenario:   b.Scenario,
		Seed:       b.Seed,
		Window:     b.Window,
		Logs:       len(b.Logs),
		Traces:     len(b.Traces),
		Signatures: len(b.Signatures),
	}

	f, err := os.Create(filepath.Join(dir, e.File))
	if err != nil {
		return Entry{}, err
	}
	if err := b.Write(f); err != nil {
		f.Close()
		return Entry{}, err
	}
	if err := f.Close(); err != nil {
		return Entry{}, err
	}

	idx.Incidents = append(idx.Incidents, e)
	return e, writeIndex(dir, idx)
}

func writeIndex(dir string, idx *Index) error {
	f, err := os.Create(filepath.Join(dir, IndexFile))
	if err != nil {
		return err
	}
	defer f.Close()
	enc := json.NewEncoder(f)
	enc.SetIndent("", "  ")
	return enc.Encode(idx)
}

File: py/diagkit_rca/history.py
"""Reader for the incident history store the Go side writes.

The store is a directory of archived bundle files plus an index.json. The Go
`diagkit archive` command appends to it; this module lists past incidents and
finds signatures that recur across them.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path

from .bundle import Bundle, load_bundle

INDEX_VERSION = "1"
INDEX_FILE = "index.json"
DEFAULT_DIR = "diagkit-history"


@dataclass
class IncidentRecord:
    id: str
    file: str
    scenario: str
    seed: int
    logs: int
    traces: int
    signatures: int


@dataclass
class Recurrence:
    template: str
    incidents: list[str] = field(default_factory=list)
    total_count: int = 0
    services: list[str] = field(default_factory=list)


def load_index(directory: str) -> list[IncidentRecord]:
    """Load the history index. A missing store is an empty history."""
    path = Path(directory) / INDEX_FILE
    if not path.exists():
        return []
    data = json.loads(path.read_text(encoding="utf-8"))
    version = data.get("version", "")
    if version != INDEX_VERSION:
        raise ValueError(
            f"unsupported history index version {version!r}, expected {INDEX_VERSION!r}"
        )
    return [
        IncidentRecord(
            id=e["id"],
            file=e["file"],
            scenario=e.get("scenario", ""),
            seed=e.get("seed", 0),
            logs=e.get("logs", 0),
            traces=e.get("traces", 0),
            signatures=e.get("signatures", 0),
        )
        for e in data.get("incidents", [])
    ]


def load_incident(directory: str, record: IncidentRecord) -> Bundle:
    return load_bundle(str(Path(directory) / record.file))


def find_recurrences(directory: str, min_incidents: int = 2) -> list[Recurrence]:
    """Signatures seen in at least min_incidents archived incidents, ranked by
    how many incidents they recur in, then by total matched log lines."""
    by_template: dict[str, Recurrence] = {}
    for record in load_index(directory):
        bundle = load_incident(directory, record)
        for sig in bundle.signatures:
            rec = by_template.setdefault(sig.template, Recurrence(template=sig.template))
            rec.incidents.append(record.id)
            rec.total_count += sig.count
            rec.services = sorted(set(rec.services) | set(sig.services))
    out = [r for r in by_template.values() if len(r.incidents) >= min_incidents]
    out.sort(key=lambda r: (-len(r.incidents), -r.total_count, r.template))
    return out
